cachix/src/Cachix/Client/Push.hs

```diff
@@ -85,7 +87,7 @@ data PushParams m r = PushParams
 data PushStrategy m r = PushStrategy
   { -- | Called when a path is already in the cache.
     onAlreadyPresent :: m r,
-    onAttempt :: RetryStatus -> Int64 -> m (),
+    onAttempt :: RetryStatus -> Int64 -> m (Maybe ProgressBar),
```

Comment:
This gives a better separation of concerns, not expanding the scope of this API client module into CLI concerns.

```suggestion
    onAttempt :: RetryStatus -> Int64 -> m (),
    -- | A conduit to interject between NAR generation and compression. This allows the caller to perform a side effect during streaming, e.g. to update a progress bar. This conduit must the NAR bytestream through unmodified.
    spyIntoNarStream :: RetryStatus -> Int64 -> ConduitM ByteString ByteString m (),
```

Besides better separation of concerns, this solution is a bit more useful because it allows the caller to perform a side effect both before and after the `yield` (which I think is meaningful wrt output blocking, but not 100% sure), and it could perhaps even be used for other purposes, such as producing a listing of the directory contents in the same pass. I don't know how relevant those use cases are, but you get them for free.